### PR TITLE
Scheduler enhancements

### PR DESCRIPTION
**Pull request by magik6k**
This is a rebased version of https://github.com/filecoin-project/lotus/pull/7269 with some cleanup

TLDR:
* Make resource accounting cgroup-aware on Linux
* Allow overriding the default resource table on a per-worker basis through env-vars
  * All envvars with defaults -> https://gist.github.com/magik6k/c0e1c7cd73c1241a9acabc30bf469a43 (generated with `lotus-worker resources --default`)
* Add proper support for multiple GPUs per worker
* Separate cpu thread counts in the resource table depending on whether or not workers have GPUs
* Added some tests to the original PR

### CODE REVIEW

File: cmd/lotus-seal-worker/resources.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	Name:  "resources",
+	Usage: "Manage resource table overrides",
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "all",
+			Usage: "print all resource envvars",
+		},
+		&cli.BoolFlag{
+			Name:  "default",
+			Usage: "print all resource envvars",
```
Comment (arajasek) [inline cmd/lotus-seal-worker/resources.go:23]:
```suggestion
			Usage: "print default resource envvars",
```

File: api/version.go [hunk 2]
```diff
@@ -58,7 +58,7 @@ var (
 	FullAPIVersion1 = newVer(2, 1, 0)
 
 	MinerAPIVersion0  = newVer(1, 2, 0)
-	WorkerAPIVersion0 = newVer(1, 1, 0)
+	WorkerAPIVersion0 = newVer(1, 5, 0)
```
[the remark this answers is not in the document]
Comment (magik6k) [reply api/version.go:61]: Testing with existing miners. Should be ok to keep like this.

**Comment by magik6k** [comment]
(rebased on latest master)

File: extern/sector-storage/sched_resources.go [hunk 3]
```diff
@@ -54,35 +54,44 @@ func (a *activeResources) free(wr storiface.WorkerResources, r Resources) {
...
+	memAvail := res.MemPhysical - memUsed
+	if memNeeded > memAvail {
+		log.Debugf("sched: not scheduling on worker %s for %s; not enough physical memory - need: %dM, have %dM available", wid, caller, memNeeded/mib, memAvail/mib)
 		return false
 	}
 
-	maxNeedMem := res.MemReserved + a.memUsedMax + needRes.MaxMemory + needRes.BaseMinMemory
+	vmemNeeded := needRes.MaxMemory + needRes.BaseMinMemory
+	vmemUsed := a.memUsedMax
+	if vmemUsed < res.MemUsed+res.MemSwapUsed {
```
Comment (ZenGround0) [inline extern/sector-storage/sched_resources.go:77]: nit: giving `res.MemUsed + res.MemSwapUsed` a name would make following along a bit easier

File: extern/sector-storage/storiface/resources.go [hunk 4]
```diff
@@ -1,19 +1,31 @@
...
 
+	"github.com/filecoin-project/go-state-types/abi"
 	"github.com/filecoin-project/lotus/extern/sector-storage/sealtasks"
 )
 
 type Resources struct {
-	MinMemory uint64 // What Must be in RAM for decent perf
-	MaxMemory uint64 // Memory required (swap + ram)
+	MinMemory uint64 `envname:"MIN_MEMORY"` // What Must be in RAM for decent perf
+	MaxMemory uint64 `envname:"MAX_MEMORY"` // Memory required (swap + ram)
```
Comment (ZenGround0) [inline extern/sector-storage/storiface/resources.go:17]: "What Must be in RAM for decent perf" makes sense to me, but "Memory required (swap + ram)" is confusing me a bit given that the name is MAX.  It sounds like this is MIN_MEMORY + MIN_SWAP?  I think clarifying this comment would be helpful.

**Review by jennijuju — APPROVED** [review]

File: extern/sector-storage/storiface/resources_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	require.NoError(t, err)
+	require.Equal(t, 2, rt[sealtasks.TTUnseal][stabi.RegisteredSealProof_StackedDrg2KiBV1_1].MaxParallelism)
+	require.Equal(t, 0.4, rt[sealtasks.TTPreCommit2][stabi.RegisteredSealProof_StackedDrg2KiBV1_1].GPUUtilization)
+	require.Equal(t, uint64(2222), rt[sealtasks.TTPreCommit2][stabi.RegisteredSealProof_StackedDrg2KiBV1_1].MaxMemory)
+
+	// check that defaults don't get mutated
+	require.Equal(t, 1, ResourceTable[sealtasks.TTUnseal][stabi.RegisteredSealProof_StackedDrg2KiBV1_1].MaxParallelism)
+}
+
+func TestListResourceSDRMulticoreOverride(t *testing.T) {
```
Comment (ZenGround0) [inline extern/sector-storage/storiface/resources_test.go:48]: Nice test

File: extern/sector-storage/storiface/resources.go [hunk 6]
```diff
@@ -323,3 +345,83 @@ func init() {
...
+
+				envname := f.Tag.Get("envname")
+				if envname == "" {
+					return nil, xerrors.Errorf("no envname for field '%s'", f.Name)
+				}
+
+				envval, found := lookup(taskType.Short()+"_"+shortSize+"_"+envname, fmt.Sprint(rr.Elem().Field(i).Interface()))
+				if !found {
+					// special multicore SDR handling
+					if (taskType == sealtasks.TTPreCommit1 || taskType == sealtasks.TTUnseal) && envname == "MAX_PARALLELISM" {
```
Comment (ZenGround0) [inline extern/sector-storage/storiface/resources.go:376]: This isn't that bad but it makes me wonder if we are hoping to deprecate `FIL_PROOFS_USE_MULTICORE_SDR`.  Other than supporting old workflows is there a reason to keep this old pattern around?

**Review by ZenGround0 — APPROVED** [review]
This is all new to me and some of the scheduling logic (canHandleRequest for example) is still a little hard for me to follow.  But overall I understand what you are doing and this looks good.

File: extern/sector-storage/cgroups_linux.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	"math"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/cgroups"
+	cgroupv2 "github.com/containerd/cgroups/v2"
+)
+
+func cgroupV2MountPoint() (string, error) {
+	f, err := os.Open("/proc/self/mountinfo")
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/cgroups_linux.go:18]: Normal users have access to this by default

File: extern/sector-storage/manager_test.go [hunk 8]
```diff
@@ -453,3 +453,123 @@ func TestReenableWorker(t *testing.T) {
...
+
+l:
+	for {
+		st := m.WorkerStats()
+		require.Len(t, st, 1)
+		for _, w := range st {
+			if w.MemUsedMax > 0 {
+				break l
+			}
+			time.Sleep(time.Millisecond)
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/manager_test.go:504]: It shouldn't (if it does for whatever reason, the test will just time out in 30min); (Normally this test takes 0.1s to run)

File: extern/sector-storage/storiface/resources.go (hunk 6 above)
Comment (magik6k) [reply extern/sector-storage/storiface/resources.go:376]: FIL_PROOFS_USE_MULTICORE_SDR is used internally inside proofs, and we don't have a better way to pass that into proofs right now.

**Pull request merged**